Python 3 support

pe-st opened the pull request:
With the help of @telemaxx the script has received Python 3 support.

After some discussions [here](#26) and [here](#24) I'd like to have some feedback for the path taken to support Python 3 alongside Python 2 (see also CONTRIBUTING.md for the reason to keep Python 2 support for now).

I plan to release this version on 2020-04-12, but I can easily postpone it depending on the discussion.

cc @telemaxx, @jedie, @bartskowron, @jamorris, @moderation, @kjkjava, @rsjrny,
@jkall, @chs8691, @chmoelders, @TheKiteRunning, @lindback

bartskowron commented on gcexport.py:
```diff
@@ -216,34 +245,58 @@ def http_req(url, post=None, headers=None):
-                keys.append(key)
+        if python3:
+            stripped_line = line.strip()
+            if stripped_line:
+                if (isinstance(stripped_line,bytes)):
+                    stripped_line = stripped_line.decode('utf8')
+                if (isinstance(comment_char,bytes)):
+                    comment_char=comment_char.decode('utf8')             
+                if not stripped_line.startswith(comment_char):
+                    #https://python-forum.io/Thread-Diff-between-Py-2-7-and-3--14749
+                    key_value = stripped_line.split(separator)
+                    key = key_value[0].strip()
+                    value = separator.join(key_value[1:]).strip().strip('"')
+                    props[key] = value
+                    if keys != None:
+                        keys.append(key)
+        else:
+            stripped_line = line.strip()
+            if stripped_line and not stripped_line.startswith(comment_char):
+                key_value = stripped_line.split(separator)
+                key = key_value[0].strip()
+                value = separator.join(key_value[1:]).strip().strip('"')
+                props[key] = value
+                if keys != None:
+                    keys.append(key)
```
What exactly is happening in this part? This block of code looks more complex than probably it should to be. I may comment some single lines but honestly, I'm pretty sure it should be refactored first or some general comment should be added on the top of it.

pe-st replied:
IIUC the special code was introduced because at the beginning of the migration to Py 3 the properties files were read as bytes instead of strings. This has been fixed some days ago and I just removed the special code, it should not be needed anymore.

bartskowron commented on gcexport.py:
```diff
@@ -590,17 +657,17 @@ def extract_device(device_dict, details, start_time_seconds, args, http_caller,
+            if 'deviceId' not in device_meta or device_id and device_id != '0':
                 device_json = http_caller(URL_GC_DEVICE + str(device_app_inst_id))
-                file_writer(args.directory + '/device_' + str(device_app_inst_id) + '.json',
+                file_writer(args.directory + sep + 'device_' + str(device_app_inst_id) + '.json',
```
```suggestion
                file_writer(join(args.directory, f'device_{device_app_inst_id}.json'),
```

bartskowron replied:
Note: I used join() as this is how you imported the function. However, it should be `os.path.join` as its more readable.

pe-st replied:
If I understand correctly (having found only https://stackoverflow.com/a/46182112/3686) it's a bit cumbersome to use f-strings with Python 2 (must install `future-fstrings` with pip etc), meaning complicating the setup instructions for systems with Python 2. Or am I missing something?

bartskowron replied:
You are right. I'm sorry, forgot the f-strings are not supported in Python2 by default.
There are few more readable options for string concatenations, however by using `+` operator is usually much quicker. As the project is not large, lets keep it as you have already proposed in the pull request.

bartskowron replied:
You are right. I'm sorry, forgot the f-strings are not supported in Python2 by default.
There are few more readable options for string concatenations, however by using + operator is usually much quicker. As the project is not large, lets keep it with the `+` operator.
However there is still way for improvements
```suggestion
                file_writer(join(args.directory, 'device_' + str(device_app_inst_id) + '.json'),
```

bartskowron commented on gcexport.py:
```diff
@@ -620,18 +687,18 @@ def extract_device(device_dict, details, start_time_seconds, args, http_caller,
         if gear:
             del args # keep 'args' argument in case you need to uncomment write_to_file
-            # write_to_file(args.directory + '/activity_' + activity_id + '-gear.json',
+            # write_to_file(args.directory + sep + 'activity_' + activity_id + '-gear.json',
```
```suggestion
            # write_to_file(join(args.directory, f'activity_{activity_id}-gear.json'),
```

bartskowron replied:
Python2 compatibility update
```suggestion
            # write_to_file(join(args.directory, 'activity_' + activity_id + '-gear.json'),
```

bartskowron commented on gcexport.py:
```diff
@@ -652,28 +719,28 @@ def export_data_file(activity_id, activity_details, args, file_time, append_desc
     # timestamp as prefix for filename
-    if args.fileprefix > 0:
-        prefix = "{}-".format(start_time_locale.replace("-", "").replace(":", b"").replace(" ", "-"))
+    if args.fileprefix and args.fileprefix > 0:
```
This looks buggy, or buggy in the future. What was a purpose to change this line from the origin? I guess there is a case when the old one provides wrong results, could you please show the case?

pe-st replied:
This change was a consequence from running `python2 -3` before the first tries with python3, I didn't really try if it's needed with python3

bartskowron replied:
Are you sure? Adding a logic with `and` has nothing to do with a compatibility. What I'm pointing here is the `args.fileprefix and`.

bartskowron commented on gcexport.py:
```diff
@@ -652,28 +719,28 @@ def export_data_file(activity_id, activity_details, args, file_time, append_desc
-    if args.fileprefix > 0:
-        prefix = "{}-".format(start_time_locale.replace("-", "").replace(":", b"").replace(" ", "-"))
+    if args.fileprefix and args.fileprefix > 0:
+        prefix = "{}-".format(start_time_locale.replace("-", "").replace(":", "").replace(" ", "-"))
```
```suggestion
        prefix = f"{start_time_locale.replace("-", "").replace(":", "").replace(" ", "-")}-"
```

bartskowron replied:
Python2 compatibility update
```suggestion
        prefix = start_time_locale.replace("-", "").replace(":", "").replace(" ", "-") + "-"
```

bartskowron replied:
This one is missing, I think.

bartskowron commented on gcexport.py:
```diff
@@ -652,28 +719,28 @@ def export_data_file(activity_id, activity_details, args, file_time, append_desc
         download_url = URL_GC_GPX_ACTIVITY + activity_id + '?full=true'
         file_mode = 'w'
     elif args.format == 'tcx':
-        data_filename = directory + '/' + prefix + 'activity_' + activity_id + append_desc + '.tcx'
+        data_filename = directory + sep + prefix + 'activity_' + activity_id + append_desc + '.tcx'
         download_url = URL_GC_TCX_ACTIVITY + activity_id + '?full=true'
         file_mode = 'w'
     elif args.format == 'original':
-        data_filename = directory + '/' + prefix + 'activity_' + activity_id + append_desc + '.zip'
+        data_filename = directory + sep + prefix + 'activity_' + activity_id + append_desc + '.zip'
         # TODO not all 'original' files are in FIT format, some are GPX or TCX...
-        fit_filename = directory + '/' + prefix + 'activity_' + activity_id + append_desc + '.fit'
+        fit_filename = directory + sep + prefix + 'activity_' + activity_id + append_desc + '.fit'
         download_url = URL_GC_ORIGINAL_ACTIVITY + activity_id
         file_mode = 'wb'
     elif args.format == 'json':
-        data_filename = directory + '/' + prefix + 'activity_' + activity_id + append_desc + '.json'
+        data_filename = directory + sep + prefix + 'activity_' + activity_id + append_desc + '.json'
```
Here is a suggestion for the first case but the other should updated in the same way
```
        data_filename = join(directory, f'{prefix}activity_{activity_id}{append_desc}.gpx')
```
Also, `data_filename` should be replace with `data_filepath` due to using a directory path at the beginning.
And missing namespacing (`os.path.join`)

pe-st replied:
In the PR #32 the need for more flexible filename patterns arose; I'd like to rewrite this code anyway

bartskowron replied:
Still it should be like this:
`data_filename = join(directory, prefix + 'activity_' + activity_id + append_desc + '.gpx')`

Other than that, I agree it should be rewritten one day ;)

bartskowron commented on gcexport.py:
```diff
@@ -732,7 +799,7 @@ def export_data_file(activity_id, activity_details, args, file_time, append_desc
                     # prepend 'activity_' and append the description to the base name
                     name_base, name_ext = splitext(name)
-                    new_name = directory + '/activity_' + name_base + append_desc + name_ext
+                    new_name = directory + sep + prefix + 'activity_' + name_base + append_desc + name_ext
```
```suggestion
                    new_name = join(directory, f'{prefix}activity_{name_base}{append_desc}{name_ext}
```

bartskowron replied:
Python2 compatibility

```suggestion
                    new_name = join(directory, prefix + 'activity_' + name_base + append_desc + name_ext)
```

bartskowron commented on gcexport.py:
```diff
@@ -765,12 +832,12 @@ def logging_verbosity(verbosity):
         if isinstance(handler, logging.FileHandler):
             # this is the logfile handler
-            level = logging.DEBUG if verbosity > 0 else logging.INFO
+            level = logging.DEBUG if verbosity and verbosity > 0 else logging.INFO
```
Why this change?

pe-st replied:
This change was a consequence from running `python2 -3` before the first tries with python3, I didn't really try if it's needed with python3

bartskowron replied:
Could you please paste output of the running `python2 -3` without this change? I truly looks weird to me.

pe-st replied:
Seems if `verbosity` isn't set on the command line it's `None` and can't be compared to 0:

    gcexport.py:835: DeprecationWarning: comparing unequal types not supported in 3.x
      level = logging.DEBUG if verbosity > 0 else logging.INFO

telemaxx replied:
we could change one line from:
`parser.add_argument('-v', '--verbosity', action='count', help='increase output verbosity')`
to:
`parser.add_argument('-v', '--verbosity', action='count', default=0, help='increase output verbosity')`

now we can remove the extra check(s):
from:
`level = logging.DEBUG if verbosity and verbosity > 0 else logging.INFO`
to
`level = logging.DEBUG if verbosity > 0 else logging.INFO`

pe-st replied:
thx @telemaxx, that simplifies the code a lot

bartskowron commented on gcexport.py:
```diff
@@ -765,12 +832,12 @@ def logging_verbosity(verbosity):
         elif isinstance(handler, logging.StreamHandler):
             # this is the console handler
-            level = logging.DEBUG if verbosity > 1 else (logging.INFO if verbosity > 0 else logging.WARN)
+            level = logging.DEBUG if verbosity and verbosity > 1 else (logging.INFO if verbosity and verbosity > 0 else logging.WARN)
```
Same question as above. Why this extra checking of `verbosity`?

bartskowron commented on gcexport.py:
```diff
@@ -844,7 +914,7 @@ def main(argv):
         print('Fetching user stats...', end='')
         logging.info('Userstats page %s', URL_GC_USERSTATS + display_name)
-        result = http_req(URL_GC_USERSTATS + display_name)
+        result = http_req_as_string(URL_GC_USERSTATS + display_name)
```
Again, I would suggest to use literal string interpolation (see PEP 498):
```suggestion
        result = http_req_as_string(f'{URL_GC_USERSTATS}display_name')
```
However, in this specific case it doesn't look that good ;)

bartskowron replied:
As for Python2 my suggestion should be dropped and keep the original change from the PR

bartskowron commented on gcexport.py:
```diff
@@ -882,8 +952,10 @@ def main(argv):
+        
         logging.info('Activity list URL %s', URL_GC_LIST + urlencode(search_params))
-        result = http_req(URL_GC_LIST + urlencode(search_params))
+        result = http_req_as_string(URL_GC_LIST + urlencode(search_params))
```
F-strings again:
```suggestion
        result = http_req_as_string(f'{URL_GC_LIST}{urlencode(search_params)})
```

bartskowron replied:
As for Python2 my suggestion should be dropped and keep the original change from the PR

bartskowron commented on gcexport.py:
```diff
@@ -903,12 +975,12 @@ def main(argv):
                 # Display which entry we're skipping.
                 print('Skipping Garmin Connect activity ', end='')
-                print('(' + str(current_index) + '/' + str(total_to_download) + ') ', end='')
+                print('(' + str(current_index) + sep + str(total_to_download) + ') ', end='')
```
```suggestion
                print(f'({current_index}{sep}{total_to_download}) ', end='')
```

bartskowron replied:
Python2 compatibility update
```suggestion
                print('(', current_index, sep, total_to_download, ') ', sep='', end='')
```

bartskowron commented on gcexport.py:
```diff
@@ -903,12 +975,12 @@ def main(argv):
                 # Display which entry we're working on.
                 print('Garmin Connect activity ', end='')
-                print('(' + str(current_index) + '/' + str(total_to_download) + ') ', end='')
+                print('(' + str(current_index) + sep + str(total_to_download) + ') ', end='')
```
```suggestion
                print(f'({current_index}{sep}{total_to_download}) ', end='')
```

bartskowron replied:
Python 2 compatibility update
```suggestion
                print('(', current_index, sep, total_to_download, ') ', sep='', end='')
```

pe-st replied:
Just curious, what is the benefit of comma-separating the arguments to `print()` here?

bartskowron replied:
Performance. Strings are immutable objects in Python, so by using concatenation (`+` operator) you actually create new objects under the hood, which requires more CPU operations and memory allocation, this also may invokes more work for garbage collector too. As passing already existed objects to print, you just pass references.

pe-st replied:
Thanks! Done (also for the other places where applicable)

bartskowron commented on gcexport.py:
```diff
@@ -920,7 +992,7 @@ def main(argv):
                 tries = MAX_TRIES
                 while tries > 0:
-                    activity_details = http_req(URL_GC_ACTIVITY + str(actvty['activityId']))
+                    activity_details = http_req_as_string(URL_GC_ACTIVITY + str(actvty['activityId']))
```
```suggestion
                    activity_details = http_req_as_string(f'{URL_GC_ACTIVITY}{actvty['activityId']}')
```

bartskowron replied:
As for Python2 my suggestion should be dropped and keep the original change from the PR

bartskowron commented on gcexport.py:
```diff
@@ -956,21 +1028,21 @@ def main(argv):
                     try:
                         # TODO implement retries here, I have observed temporary failures
-                        activity_measurements = http_req(URL_GC_ACTIVITY + str(actvty['activityId']) + "/details")
+                        activity_measurements = http_req_as_string(URL_GC_ACTIVITY + str(actvty['activityId']) + "/details")
```
```suggestion
                        activity_measurements = http_req_as_string(f'{URL_GC_ACTIVITY}{actvty['activityId']}/details')
```

bartskowron replied:
As for Python2 my suggestion should be dropped and keep the original change from the PR

bartskowron commented on gcexport.py:
```diff
@@ -25,28 +26,45 @@
 from math import floor
 from os import makedirs, mkdir, rename, remove, stat, utime
-from os.path import dirname, isdir, isfile, join, realpath, splitext
+from os.path import dirname, isdir, isfile, join, realpath, sep, splitext
```
I would go with a proper namespacing `import os.path`

telemaxx commented on gcexport.py:
```diff
@@ -172,14 +191,19 @@ def sanitize_filename(name, max_length=0):
 def write_to_file(filename, content, mode, file_time=None):
     """Helper function that persists content to file."""
-    write_file = open(filename, mode)
+    if python3 and (filename.endswith('.json') or filename.endswith('.gpx') or filename.endswith('.tcx')):
```
cool! i never used this statement.

telemaxx commented on README.md:
```diff
@@ -91,17 +91,22 @@ optional arguments:
-`python gcexport.py --count all` will download all of your data to a dated directory.
 
-`python gcexport.py -c all -f gpx -ot --desc 20` will export all of your data in GPX format, set the timestamp of the GPX files to the start time of the activity and append the 20 first characters of the activity's description to the file name.
+- `python gcexport.py --count all`  
```
good point

pe-st replied:
Done. Thanks for pointing to the PEP!

pe-st commented:
> Hi,
> Thanks for all the efforts! I'm leaving some comments and suggestions.
> 
> I also suggest to divide python2 and python3 compatible codes in distinct modules with general runner for endusers. This way it would be a way easier to maintain and drop py2 one day too. I can rewrite it to this stage if you wish to, once the current pull request will be approved.
> 
> I don't also the distinction regarding Unicode. I believe it would be better to normalise it for both versions. This required some extra work though. Here is some explanation: https://stackoverflow.com/questions/41542613/utf-8-string-in-python-2-and-3

Thanks! It will take some days to properly address all these issues ;-)

bartskowron commented:
Updated some suggestions to avoid the using of f-strings.

However, some of them should be kept as in original proposal in the PR. I didn't remove the original comments/suggestions to not making additional mess. I don't have write access to close them, so just added extra comments that they are not compatible with the Python 2. Please close them.

pe-st commented:
> @pe-st Thanks for adding me to the contributors list. However, I would suggest to use "commit suggestion" feature, so the proposed and accepted changes will be committed automatically by GitHub with proper credits :)
> EDIT: This will also be better handled by GitHub to see which changes are done, not outdated. Now its a bit messy as I have to recheck all changes :(

Sorry about that. Next time I'll add the `Co-authored-by:` line to my commit message for proper attribution. I suppose that when you use the provided button you get a separate commit for every suggestion, which IMHO clutters the commit history...

bartskowron commented:
IMHO commits should be split by functionality and only by functionality. It doesn't really matter if you have 10000 or 100 commits in the history. What matters is to easily find a specific commit and have a possibility to roll it back without complications in other parts of the code. Regarding my suggestions, some should be committed in groups (e.g. all refactoring of replacing concat by `os.path.join()` or changes in README regarding running `python3` interpreter) but never as one commit.

I definitely agree with applying and testing suggestions locally first. Rule of thumb is to **always** test before commit ;)

Speaking of the `Co-authored-by:` looks like my privacy settings were set wrongly and my email address is exposed in the commit message now :( I updated my settings on GitHub to get a private email. Any chance to fix it for this particular commit?

pe-st commented:
> IMHO commits should be split by functionality and only by functionality. It doesn't really matter if you have 10000 or 100 commits in the history. What matters is to easily find a specific commit and have a possibility to roll it back without complications in other parts of the code. Regarding my suggestions, some should be committed in groups (e.g. all refactoring of replacing concat by `os.path.join()` or changes in README regarding running `python3` interpreter) but never as one commit.

Absolutely.

> Speaking of the `Co-authored-by:` looks like my privacy settings were set wrongly and my email address is exposed in the commit message now :( I updated my settings on GitHub to get a private email. Any chance to fix it for this particular commit?

Done, I replaced the commit.

bartskowron commented:
> > Speaking of the `Co-authored-by:` looks like my privacy settings were set wrongly and my email address is exposed in the commit message now :( I updated my settings on GitHub to get a private email. Any chance to fix it for this particular commit?
> Done, I replaced the commit.

Thank you! I really appreciate it as I know it wasn't the easiest job ;)

pe-st commented:
> I plan to release this version on 2020-04-12, but I can easily postpone it depending on the discussion.

With all the valid points still open I hope to release it one week later on 2020-04-19

bartskowron commented:
Is there still any ongoing issues that you plan to work on before the release? Github seems to show me most as resolved, however I remember about some aspects that I haven't seen any changes nor comments. @pe-st please let me know if you are still working on something here or I could run another round of a review now.

pe-st commented:
> Is there still any ongoing issues that you plan to work on before the release? Github seems to show me most as resolved, however I remember about some aspects that I haven't seen any changes nor comments. @pe-st please let me know if you are still working on something here or I could run another round of a review now.

I still want to have a look at the `write_to_file` function before closing the PR.

The unicode normalisation I'll have to study probably at a later stage, considering my time budget ;-)

pe-st commented:
> > Is there still any ongoing issues that you plan to work on before the release? Github seems to show me most as resolved, however I remember about some aspects that I haven't seen any changes nor comments. @pe-st please let me know if you are still working on something here or I could run another round of a review now.
> 
> I still want to have a look at the `write_to_file` function before closing the PR.

@bartskowron if there are no big gotchas left I'll release the current version in a couple of days. Further improvements can then be applied to future versions, this PR has been open long enough ;-)

bartskowron commented on gcexport.py:
```diff
@@ -197,22 +198,37 @@ def sanitize_filename(name, max_length=0):
+        write_file = io.open(filename, mode, encoding="utf-8")
+        if isinstance(content, bytes):
+            content = content.decode("utf-8")
+    else: # 'wb'
+        write_file = io.open(filename, mode)
```
Using `else` as assumption that it will be 'wb' can be very wrong. I would suggest a better handling of errors like this.

```suggestion
    elif mode=='wb':
        write_file = io.open(filename, mode)
    else:
        raise Exception('Unsupported file mode:' mode)
```

pe-st replied:
I agree

bartskowron commented on gcexport.py:
```diff
@@ -197,22 +198,37 @@ def sanitize_filename(name, max_length=0):
 
 
 def write_to_file(filename, content, mode, file_time=None):
```
IMHO its good idea to make a default value here. Especially as by reading the docstrings user cant be sure which one to pick.

```suggestion
def write_to_file(filename, content, mode='w', file_time=None):
```

pe-st commented on gcexport.py:
```diff
@@ -24,29 +25,47 @@
+    import urllib2
+    from urllib import urlencode
+    COOKIE_JAR = cookielib.CookieJar()
+    OPENER = urllib2.build_opener(urllib2.HTTPCookieProcessor(COOKIE_JAR), urllib2.HTTPSHandler(debuglevel=0))
```
Yes, this is cleaner

pe-st commented:
> Some minor and last suggestions :)

https://github.com/pe-st/garmin-connect-export/pull/33#discussion_r416852513

Are you referencing the test of `args.fileprefix` against `None`? This has been taken care with a default value as suggested by @telemaxx , see commit f8e60847

bartskowron commented:
> [#33 (comment)](https://github.com/pe-st/garmin-connect-export/pull/33#discussion_r416852513)
> 
> Are you referencing the test of `args.fileprefix` against `None`? This has been taken care with a default value as suggested by @telemaxx , see commit [f8e6084](https://github.com/pe-st/garmin-connect-export/commit/f8e608472911f917ce27ab42fcc1523a4ac48830)

No, I was referencing to the string formatting but honestly it looks correct now. Maybe I misread something before.

It's all right now :)
